students/mkdir01/lesson03/mailroom.py

```diff
@@ -0,0 +1,106 @@
+        name = input("Please enter a full name (or type \'list\', or hit \'return\' to go back)> ")
+        if name == "list":
+            for row in data:
+                print(row[0])  # name is always the 1st entry in each data row
+                sub = 0  # runs the subroutine again
+        elif name == "":
+            sub = 0
+            continue
+        else:
+            for i, row in enumerate(data):  # search for a name match
+                if row[0] == name:
+                    new_row = data.pop(i)  # if name is found pop() entry and put in new_row
+                    break
+            else:  # if no name match, make new entry
+                new_row = [name]
+            dollars = input("Please enter a donation amount (ex: 500.05)> ")
+            new_row.append(round(float(dollars), 2))  # b/c I don't understand floating point math
+            data.append(new_row)
+            print(f"Dear {new_row[0]},\n\nThank you for your donation of {new_row[-1]:.2f}.\n\nLove,\nThe Charity Org\n")
+        sub = not sub  # leaves subroutine, unless 'list' has been called
+    else:
+        return
+
+
+# mode 2 - "create a report"
+def create_report():
+    # create temp data
+    temp_data = []  # resets temp data
+    for row in data:
+        temp_row = []  # reset temp_row
+        temp_total = 0.00  # reset temp_total
+        for cell in row:
+            if not temp_row:  # first cell - when temp_row is empty
+                temp_row.append(cell)  # append donor name, name is always in first cell
+            else:
+                temp_total += float(cell)  # total gifts, added from cells after first
+        temp_row.append(round(float(temp_total), 2))  # append historical total
+        temp_row.append(len(row) - 1)  # append number of gifts
+        temp_row.append(round(temp_total / (len(row) - 1), 2))  # append avg gift
+        temp_data.append(temp_row)  # adds row to temp data structure, will build new data structure by end of for loop
+
+    # sort temp data
+    # counter will ensure we've taken one full pass of the data where all rows are in order before quitting
+    counter = 0
+    temp_row = []
+    while counter < (len(temp_data) - 1):  # we will get an out of bounds error if we don't stop at len-1
+        counter = 0
+        for i in range(len(temp_data) - 1):  # I think this is a bubble sort? It's been a while...
+            if temp_data[i][1] < temp_data[i + 1][1]:  # if this row's total is less than the next row's total...
+                temp_row = temp_data.pop(i)  # pop() the row...
+                temp_data.insert(i + 1, temp_row)  # insert() the row into the next cell
+            else:
+                counter += 1  # when we get through a whole for loop withought "bubbling" we leave the while loop
+                # there's probably a more elegent way to make sure we've fully bubbled, but I don't know it
```

Comment:
you can simplify rows 39 to 64.

you need one for loop to create summary list: -> `[[<name>, <avg>, <total>, <sum>]]`
then sort the data using `sorted` function, no need to reimplement sort
https://wiki.python.org/moin/HowTo/Sorting